lightning/restore/restore.go

```diff
@@ -1332,6 +1345,91 @@ func (rc *RestoreController) switchTiKVMode(ctx context.Context, mode sstpb.Swit
 	)
 }
 
+func (rc *RestoreController) enforceDiskQuota(ctx context.Context) {
+	if !atomic.CompareAndSwapInt32(&rc.diskQuotaState, diskQuotaStateIdle, diskQuotaStateChecking) {
+		// do not run multiple the disk quota check / import simultaneously.
+		// (we execute the lock check in background to avoid blocking the cron thread)
+		return
+	}
+
+	go func() {
+		// locker is assigned when we detect the disk quota is exceeded.
+		// before the disk quota is confirmed exceeded, we keep the diskQuotaLock
+		// unlocked to avoid periodically interrupting the writer threads.
+		var locker sync.Locker
+		defer func() {
+			atomic.StoreInt32(&rc.diskQuotaState, diskQuotaStateIdle)
+			if locker != nil {
+				locker.Unlock()
+			}
+		}()
+
+		isRetrying := false
+
+		for {
+			// sleep for a cycle if we are retrying because there is nothing new to import.
+			if isRetrying {
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(rc.cfg.Cron.CheckDiskQuota.Duration):
+				}
+			} else {
+				isRetrying = true
+			}
+
+			quota := int64(rc.cfg.TikvImporter.DiskQuota)
+			largeEngines, inProgressLargeEngines, totalDiskSize, totalMemSize := rc.backend.CheckDiskQuota(quota)
+			metric.LocalStorageUsageBytesGauge.WithLabelValues("disk").Set(float64(totalDiskSize))
+			metric.LocalStorageUsageBytesGauge.WithLabelValues("mem").Set(float64(totalMemSize))
+
+			logger := log.With(
+				zap.Int64("diskSize", totalDiskSize),
+				zap.Int64("memSize", totalMemSize),
+				zap.Int64("quota", quota),
+				zap.Int("largeEnginesCount", len(largeEngines)),
+				zap.Int("inProgressLargeEnginesCount", inProgressLargeEngines))
+
+			if len(largeEngines) == 0 && inProgressLargeEngines == 0 {
+				logger.Debug("disk quota respected")
+				return
+			}
+
+			if locker == nil {
+				// blocks all writers when we detected disk quota being exceeded.
+				rc.diskQuotaLock.Lock()
+				locker = &rc.diskQuotaLock
+			}
+
+			logger.Warn("disk quota exceeded")
+			if len(largeEngines) == 0 {
+				logger.Warn("all large engines are already importing, keep blocking all writes")
+				continue
+			}
+
+			// flush all engines so that checkpoints can be updated.
+			if err := rc.backend.FlushAll(); err != nil {
+				logger.Error("flush engine for disk quota failed, check again later", log.ShortError(err))
+				return
+			}
+
+			// at this point, all engines are synchronized on disk.
+			// we then import every large engines one by one and complete.
+			// if any engine failed to import, we just try again next time, since the data are still intact.
+			atomic.StoreInt32(&rc.diskQuotaState, diskQuotaStateImporting)
+			task := logger.Begin(zap.WarnLevel, "importing large engines for disk quota")
+			var importErr error
+			for _, engine := range largeEngines {
+				if err := rc.backend.UnsafeImportAndReset(ctx, engine); err != nil {
+					importErr = multierr.Append(importErr, err)
+				}
+			}
+			task.End(zap.ErrorLevel, importErr)
```

Comment:
What if some or all engines imported failed here, e.g. by splitting region failures or one or more tikv instances are down. In the current implementation, this function will return with an error, and all write will then resume while the disk is nearly full before the next `checkDiskQuota` is scheduled. I'm afraid in some conditions, the pebble routine will be stuck at flush when the disk is full as we have met.

Reply:
🤔 no good idea yet.

things we know won't work:

* retry import on error - we already retried internally, more retry won't help
* use a state to record if the engine has been flushed to avoid double-flushing - won't work because after the mutex is unlocked, some more rows will be written to make the engine dirty again
* just skip 1 minute (replace the `return` next line by `continue`) - almost same as retry, and we still double-flush, even without double-flush we can still be stuck

Reply:
It seems that we do not change the checkpoint after flushing engines? So why we must flush them before importing?